### DISCUSSION TITLE
Textual CPU usage

### DISCUSSION THREAD

**Discussion by nschloe** [Textualize/textual#172] [General]
When running tiptop and updating every 2 seconds, it eats 7% of CPU:

![screenshot](https://user-images.githubusercontent.com/181628/142831706-0d87fdc1-7c2e-4754-a913-a17051edca62.png)

It's not tiptop's own code that is causing this (which is essentially just a bunch of psutil calls every 2 seconds), but I'm not sure either what in Textual uses so much CPU. Perhaps the constant redraws?

**Comment by willmcgugan** [comment]
It's a combination of the rendering within the Python code and the work the terminal itself is doing.

Tiptop must be doing some of that work though. Those beautiful braille graph widgets must require some cpu.

The tables are expensive as they were built to be extremely flexible. There's a lot of work done to calculate optimal columns and cell sizes. In the future, Textual will have a dynamic datatable widget which won't have quite so many options, but renders much quicker. Future versions will also get faster as hotspots are optimized.

**Reply by nschloe** [reply]
> Those beautiful braille graph widgets must require some cpu.

It's only a multiline string, and every two seconds, the last column is updated. If anything, it's the string redraws, but I'd say even those are cheap.

> The tables are expensive as they were built to be extremely flexible. There's a lot of work done to calculate optimal columns and cell sizes.

I understand. Since the tiptop layout is static, it'd probably help to have updatable table content, instead of having to recreate the table layout on every render.

**Comment by willmcgugan** [comment]
There are some experimental optimizations in Rich 10.15.0a0 which *may* make a difference.

**Reply by nschloe** [reply]
It's not noticable.

**Reply by willmcgugan** [reply]
It would be a useful exercise to profile it, to see where the hotspots are.

**Reply by nschloe** [reply]
Absolutely! Any hints on how to profile textual apps?

**Reply by willmcgugan** [reply]
It shouldn't be much different from any other app. There is cProfile in the stdlib. Another profiler I've used is "scalene". TBH I've never found a really great profiler for Python, but there should be more out there.

**Reply by tusharsadhwani** [reply]
@willmcgugan try out https://pypi.org/p/yelp-gprof2dot

**Reply by willmcgugan** [reply]
I would expect a graph visualisation would be way to large to be helpful!

**Reply by nschloe** [reply]
I wrote [tuna](https://github.com/nschloe/tuna) for cProfile visualization, but that's not the problem here. It's basically impossible to profile [tiptop](https://github.com/nschloe/tiptop) with
```
python3 -mcProfile -o program.prof ​/home/nschloe/.local/bin/tiptop
```
Even after running for one second, it won't exit gracefully. `kill`ing is the only option, and that doesn't leave a profile. What's more, after killing every mouse movement produces those dreaded

![screenshot](https://user-images.githubusercontent.com/181628/143273055-39e8b7ce-c9d0-4041-9220-eb8a14975bb9.png)

patterns.

**Reply by willmcgugan** [reply]
Hmm. Can't imagine why profiling breaks things. I'm afraid I don't have time to investigate at the moment.

**Reply by nschloe** [reply]
`cProfile`ing on a simpler version shows that most of the runtime is spent in `run()` -- understandable, that's what it's meant to do.

![screenshot](https://user-images.githubusercontent.com/181628/143320882-d1178a64-9295-4c63-ac07-67d77440c995.png)

Not sure how else to profile Textual code.

**Comment by muideen1990** [comment]
Nico Schlömer
cProfileing on a simpler version shows that most of the runtime is spent in run() -- understandable, that's what it's meant to do. Not sure how else to profile Textual code.
It look nice

[Nishat Aktar - Chat @ Spike](https://spikenow.com/r/a/?ref=spike-organic-signature&_ts=1b3a3s)	[1b3a3s]

On November 24, 2021 at 22:32 GMT, Nico Schlömer ***@***.***> wrote:

cProfileing on a simpler version shows that most of the runtime is spent in run() -- understandable, that's what it's meant to do.

[screenshot](https://user-images.githubusercontent.com/181628/143320882-d1178a64-9295-4c63-ac07-67d77440c995.png)

Not sure how else to profile Textual code.
